services/leanhelixterm/leanhelix_term.go

```diff
@@ -35,41 +39,66 @@ func NewLeanHelixTerm(ctx context.Context, logger logger.LHLogger, config *inter
-	logger.ConsensusTrace("got committee for the current consensus round", nil, log.StringableSlice("committee", committeeMembers))
+	// on ctx terminated requestOrderedCommitteePersist returns nil committee
+	isParticipating := isParticipatingInTerm(myMemberId, committeeMembers)
 
 	if !isParticipating {
-		logger.Info("OUT OF COMMITTEE: H=%d, prevBlockProof=%s, randomSeed=%d, members=%s, isParticipating=%t", blockHeight, printShortBlockProofBytes(prevBlockProofBytes), randomSeed, termincommittee.ToCommitteeMembersStr(committeeMembers), isParticipating)
+		logger.Debug("OUT OF COMMITTEE: H=%d, prevBlockProof=%s, randomSeed=%d, members=%s, isParticipating=%t", blockHeight, printShortBlockProofBytes(prevBlockProofBytes), randomSeed, termincommittee.ToCommitteeMembersStr(committeeMembers), isParticipating)
 		return termNotInCommittee(randomSeed, config)
 	}
 
+	logger.Debug("RECEIVED COMMITTEE: H=%d, prevBlockProof=%s, randomSeed=%d, members=%s, isParticipating=%t", blockHeight, printShortBlockProofBytes(prevBlockProofBytes), randomSeed, termincommittee.ToCommitteeMembersStr(committeeMembers), isParticipating)
+	logger.ConsensusTrace("got committee for the current consensus round", nil, log.StringableSlice("committee", committeeMembers))
+
 	termInCommittee := termincommittee.NewTermInCommittee(logger, config, state, messageFactory, electionTrigger, committeeMembers, prevBlock, canBeFirstLeader, CommitsToProof(logger, config.KeyManager, onCommit))
 	return &LeanHelixTerm{
 		ConsensusMessagesFilter: NewConsensusMessagesFilter(termInCommittee, config.KeyManager, randomSeed),
 		termInCommittee:         termInCommittee,
 	}
 }
 
-func requestOrderedCommittee(s *state.State, blockHeight primitives.BlockHeight, randomSeed uint64, config *interfaces.Config) ([]primitives.MemberId, error) {
+func requestOrderedCommitteePersist(s *state.State, blockHeight primitives.BlockHeight, randomSeed uint64, config *interfaces.Config, logger logger.LHLogger) ([]primitives.MemberId, error) {
 	const maxView = primitives.View(math.MaxUint64)
 	ctx, err := s.Contexts.For(state.NewHeightView(blockHeight, maxView)) // term-level context
 	if err != nil {
 		return nil, err
 	}
-	committeeMembers, err := config.Membership.RequestOrderedCommittee(ctx, blockHeight, randomSeed)
-	if err != nil {
-		return nil, err
+	logger.Debug("Polling RequestOrderedCommittee: H=%d, interval-between-attempts=%d", blockHeight, CallCommitteeContractInterval)
+
+	attempts := 1
+	for {
+
+		// exit on term update (node sync) or system shutdown
+		if ctx.Err() != nil {
+			return nil, errors.Wrap(ctx.Err(), "requestOrderedCommitteePersist: context terminated")
+		}
+
+		committeeMembers, err := config.Membership.RequestOrderedCommittee(ctx, blockHeight, randomSeed)
+		if err == nil {
+			return committeeMembers, nil
+		}
+
```

Comment:
Is there a way to tell if this error is recoverable (i.e further attempts may succeed)? If it isn't, there is a danger of infinite loop here (at least until the context is cancelled).

Reply:
This code was transferred from the consensusContext - RequestOrderingCommittee. 
Leaving the polling only as concern of the consumer (leanHelix).
The endless loop inside consensusContext is the bug found which "killed" the node sync single thread.
Since there are cases where waiting can prove viable and will conclude in participating in the term it is a reasonable behavior inside leanHelix.
LeanHelix remains responsive in such a case with its main signaling thread on listen mode.
Since leanHelix cannot participate in a term unless it is part of the committee it waits until a new term signal will arrive from main loop - via ctxt cancel.
The polling can be "optimized\cleaned" for specific errors but it is not very interesting as it will result in the same underlying behavior - waiting out for the main loop to signal the next new term (with ctxt cancel)